File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <iostream>
#include <queue>
using namespace std;

int dx[4]={0,0,1,-1};
int dy[4]={1,-1,0,0};
int main(){
  int t;
  cin>>t;
  for(int z=0;z<t;z++){
    int cnt=0;
    int arr[51][51]={0};
    int m,n,k;
    cin>>m>>n>>k;
    for(int i=0;i<k;i++){
      int x,y;
      cin>>y>>x;
      arr[x][y]=1;
    }

    // for(int i=0;i<n;i++){
    //   for(int j=0;j<m;j++){
    //     cout<<arr[i][j]<<" ";
    //   }
    //   cout<<endl;
    // }
    
    for(int i=0;i<n;i++){
      for(int j=0;j<m;j++){
        if(arr[i][j]==1){
          queue<pair<int,int>> q;
          if(arr[i][j]!=0){
            cnt++;
            q.push({i,j});
            arr[i][j]=0;
          }
          while(!q.empty()){
            int start=q.front().first;
            int end=q.front().second;
            q.pop();

            for(int w=0;w<4;w++){
              int nx=start+dx[w];
              int ny=end+dy[w];

              if(nx>=0 && ny>=0 && nx<n && ny<m && arr[nx][ny]==1){
                q.push({nx, ny});
                arr[nx][ny]=0;
              }
            }
          }
        }
      }
    }
    // for(int i=0;i<n;i++){
    //   for(int j=0;j<m;j++){
    //     cout<<arr[i][j]<<" ";
    //   }
    //   cout<<endl;
    // }
    cout<<cnt<<endl;
  }
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {
	public static int[] dx = {1,0,0,-1};
	public static int[] dy = {0,1,-1,0};
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int num = Integer.parseInt(br.readLine());
		
		for(int i=0;i<num;i++) {
			int cnt = 0;
			Queue<Pair> q = new LinkedList<Pair>();
			String[] docs = br.readLine().split(" ");
			int m = Integer.parseInt(docs[0]);
			int n = Integer.parseInt(docs[1]);
			int l = Integer.parseInt(docs[2]);
			int[][] arr = new int[n][m];
			boolean[][] visited = new boolean[n][m];
			for(int j=0;j<l;j++) {
				String[] tmp = br.readLine().split(" ");
				int x = Integer.parseInt(tmp[0]);
				int y = Integer.parseInt(tmp[1]);
				arr[y][x] = 1;
			}
			for(int j=0;j<n;j++) {
				for(int k=0;k<m;k++) {
					if(arr[j][k]==1 && !visited[j][k]) {
						q.add(new Pair(j,k));
						visited[j][k] = true;
						while(!q.isEmpty()) {
							Pair p = q.poll();
							int x = p.x;
							int y = p.y;
							for(int z=0;z<4;z++) {
								int nx = x+dx[z];
								int ny = y+dy[z];
								if(nx>=0 && ny>=0 && nx<n && ny<m && !visited[nx][ny] && arr[nx][ny]==1) {
									visited[nx][ny] = true;
									q.add(new Pair(nx,ny));
								}
							}
						}
						cnt++;
					}
				}
			}
			System.out.println(cnt);
		}
	}
	
	public static class Pair{
		int x;
		int y;
		Pair(int x, int y){
			this.x = x;
			this.y = y;
		}
	}
}
